flair/models/biomedical_entity_linking.py

```diff
@@ -0,0 +1,988 @@
+        dict_embeds: np.ndarray,
+        cosine: bool = False,
+        normalise: bool = False,
+    ) -> np.ndarray:
+        """
+        Return score matrix
+        :param query_embeds: 2d numpy array of query embeddings
+        :param dict_embeds: 2d numpy array of query embeddings
+        :param score_matrix: 2d numpy array of scores
+        """
+        if cosine:
+            score_matrix = cosine_similarity(query_embeds, dict_embeds)
+        else:
+            score_matrix = np.matmul(query_embeds, dict_embeds.T)
+
+        if normalise:
+            score_matrix = (score_matrix - score_matrix.min()) / (
+                score_matrix.max() - score_matrix.min()
+            )
+
+        return score_matrix
+
+    def retrieve_sparse_topk_candidates(
+        self,
+        score_matrix: np.ndarray,
+        topk: int,
+        batch_size: int = 128,
+        show_progress: bool = False,
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Return sorted topk idxes (descending order)
+        :param score_matrix: 2d numpy array of scores
+        :param topk: number of candidates to retrieve
+        :return res: d numpy array of ids [# of query , # of dict]
+        :return scores: numpy array of top scores
+        """
+
+        res = None
+        scores = None
+        for i in tqdm(
+            np.arange(0, score_matrix.shape[0], batch_size), disable=not show_progress
+        ):
+            score_matrix_tmp = torch.tensor(
+                score_matrix[i : i + batch_size], device=flair.device
+            )
+            sorted_values, matrix_sorted = torch.sort(
+                score_matrix_tmp, dim=1, descending=True
+            )
+            matrix_sorted = matrix_sorted[:, :topk].cpu()
+            sorted_values = sorted_values[:, :topk].cpu()
+            if res is None:
+                res = matrix_sorted
+                scores = sorted_values
+            else:
+                res = torch.cat([res, matrix_sorted], axis=0)
+                scores = torch.cat([scores, sorted_values], axis=0)
+
+        return res.numpy(), scores.numpy()
+
+    def get_predictions(
+        self,
+        mention: str,
+        topk: int,
+        tgt_space_mean_vec: np.ndarray = None,
+        batch_size: int = 1024,
+    ) -> np.ndarray:
+        """
+        Return the topk predictions for a mention and their scores
+        :param mention: string of the mention to find candidates for
+        :param topk: number of candidates
+        :return res: d numpy array of ids [# of query , # of dict]
+        :return scores: numpy array of top predictions and their scores
+        """
+        # get dense embeds for mention
+        mention_dense_embeds = self.embed_dense(names=[mention])
+
+        # remove mean centering
+        if tgt_space_mean_vec is not None:
+            mention_dense_embeds -= tgt_space_mean_vec
+
+        # normalize mention vector
+        faiss.normalize_L2(mention_dense_embeds)
```

Comment:
We do not need L2 normalization unless we are using cosine similarity (see [here](https://github.com/facebookresearch/faiss/wiki/MetricType-and-distances#how-can-i-index-vectors-for-cosine-similarity))